extensions/cache/runtime/src/main/java/io/quarkus/cache/runtime/CacheInterceptor.java

```diff
@@ -0,0 +1,88 @@
+import javax.interceptor.Interceptor;
+import javax.interceptor.InvocationContext;
+
+import org.jboss.logging.Logger;
+
+import io.quarkus.cache.CacheInvalidate;
+import io.quarkus.cache.CacheInvalidateAll;
+import io.quarkus.cache.CacheKey;
+import io.quarkus.cache.CacheResult;
+import io.quarkus.cache.runtime.caffeine.CaffeineCache;
+
+@CacheInterceptorBinding
+@Interceptor
+@Priority(PLATFORM_BEFORE)
+public class CacheInterceptor {
+
+    private static final Logger LOGGER = Logger.getLogger(CacheInterceptor.class);
+
+    @Inject
+    CacheRepository cacheRepository;
+
+    @AroundInvoke
+    public Object intercept(InvocationContext context) throws Exception {
+        Method method = context.getMethod();
+
+        for (CacheInvalidateAll cacheInvalidateAll : method.getDeclaredAnnotationsByType(CacheInvalidateAll.class)) {
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debugf("Invalidating all entries from cache [%s]", cacheInvalidateAll.cacheName());
+            }
+            CaffeineCache cache = cacheRepository.getCache(cacheInvalidateAll.cacheName());
+            cache.invalidateAll();
+        }
+
+        for (CacheInvalidate cacheInvalidate : method.getDeclaredAnnotationsByType(CacheInvalidate.class)) {
+            Object key = getCacheKey(context, cacheInvalidate.cacheName());
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debugf("Invalidating entry with key [%s] from cache [%s]", key, cacheInvalidate.cacheName());
+            }
+            CaffeineCache cache = cacheRepository.getCache(cacheInvalidate.cacheName());
+            cache.invalidate(key);
+        }
+
+        CacheResult cacheResult = method.getDeclaredAnnotation(CacheResult.class);
+        if (cacheResult != null) {
+            Object key = getCacheKey(context, cacheResult.cacheName());
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debugf("Loading entry with key [%s] from cache [%s]", key, cacheResult.cacheName());
+            }
+            CaffeineCache cache = cacheRepository.getCache(cacheResult.cacheName());
+            return cache.get(key, () -> context.proceed(), cacheResult.lockTimeout());
+        }
+
+        return context.proceed();
+    }
+
+    private Object getCacheKey(InvocationContext context, String cacheName) {
+        // If the method doesn't have any parameter, then a unique default key is generated and used.
+        if (context.getParameters().length == 0) {
+            return CacheKeyBuilder.buildDefault(cacheName);
+        } else {
+            List<Object> cacheKeyElements = new ArrayList<>();
+            // If at least one of the method parameters is annotated with @CacheKey, then the key is composed of all
+            // @CacheKey-annotated parameters.
+            for (int i = 0; i < context.getParameters().length; i++) {
+                if (context.getMethod().getParameters()[i].isAnnotationPresent(CacheKey.class)) {
```

Comment:
Same here, I'm pretty sure it will end up calling `declaredAnnotations()` again.